airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/ConfigPersistenceFactory.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Airbyte
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.config.persistence;
+
+import static io.airbyte.config.persistence.AirbyteConfigsTable.AIRBYTE_CONFIGS_TABLE_SCHEMA;
+
+import io.airbyte.commons.resources.MoreResources;
+import io.airbyte.config.Configs;
+import io.airbyte.db.Database;
+import io.airbyte.db.Databases;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.function.Function;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * By default, this factory returns a database config persistence. it can still return a file system
+ * config persistence for testing purpose. This legacy feature should be removed after the file to
+ * database migration is completely done.
+ */
+public class ConfigPersistenceFactory {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ConfigPersistenceFactory.class);
+
+  private final Configs configs;
+  private final boolean setupDatabase;
+  private final boolean useConfigDatabase;
+
+  /**
+   * @param setupDatabase initialize the database and load data; this is necessary because this method
+   *        has multiple callers, and we want to setup the database only once to prevent race
+   *        conditions.
+   */
+  private ConfigPersistenceFactory(Configs configs, boolean setupDatabase, boolean useConfigDatabase) {
+    this.configs = configs;
+    this.setupDatabase = setupDatabase;
+    this.useConfigDatabase = useConfigDatabase;
+  }
+
+  public static Builder build(Configs configs) {
```

Comment:
i think there's a weird thing with naming going on here. I think the method on line 87 should be `build` (builders build stuff) instead of `get`. This method (line 63) should probably not be called `build` since it is really a static factory. maybe call it `builder`?

Reply:
I'm wondering if the Factory needs to be exposed outside of this class? I think I understand why you want it to have a testable unit, but it seems like whenever we are using `ConfigPersistenceFactory` we just immediately call create. It's not dependency injected or anything? Does the caller even need to know about it?

I put up a suggestion in a quick PR to give you and idea of what I was thinking. I think hiding the factory still allow for module testing (it also allows just testing the hard business logic of the factory as opposed to having to test the factory and builder together) and is a bit easier to call.

suggestion: https://github.com/airbytehq/airbyte/pull/4772